### PR TITLE
support for keras models part 2 - utils

### PR DESCRIPTION
**Pull request by hstojic**
support for neural nets with Keras comes in four parts:

1. interface (done)
2. utilities (this PR)
3. model and architectures (deep ensembles only)
4. trajectory sampler 
5. integration tests and notebook

### CODE REVIEW

File: trieste/models/keras/utils.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    :return: Tensor specification objects for the ``query_points`` and ``observations`` tensors.
+    :raise ValueError: If the dataset is not an instance of :class:`~trieste.data.Dataset`.
+    """
+    if not isinstance(dataset, Dataset):
+        raise ValueError(
+            f"This function works only on trieste.data.Dataset objects, however got"
+            f"{type(dataset)} which is incompatible."
+        )
+    input_tensor_spec = tf.TensorSpec(
+        shape=(dataset.query_points.shape[-1],),
```
Comment (vpicheny) [inline trieste/models/keras/utils.py]: why the last dimension only?

File: trieste/models/keras/utils.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        )
+    tf.debugging.assert_positive(len(dataset), message="Dataset must not be empty.")
+
+    n_rows = dataset.observations.shape[0]
+
+    index_tensor = tf.random.uniform(
+        (n_rows,), maxval=n_rows, dtype=tf.dtypes.int32
+    )  # pylint: disable=all
+
+    observations = tf.gather(dataset.observations, index_tensor)  # pylint: disable=all
```
Comment (vpicheny) [inline trieste/models/keras/utils.py]: don't we need an `axis=-1` here? could be useful if we want this to work with rank > 2 tensors?

File: tests/unit/models/keras/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    with pytest.raises(tf.errors.InvalidArgumentError):
+        sample_with_replacement(dataset)
+
+
+@random_seed
+def test_sample_with_replacement_seems_correct() -> None:
+
+    n_rows = 100
+    x = tf.constant(np.arange(0, n_rows, 1), shape=[n_rows, 1])
```
Comment (vpicheny) [inline tests/unit/models/keras/test_utils.py]: worth checking with x in dim higher than 1? (see other comment below)

**Review by vpicheny — APPROVED** [review]
LGTM - left a couple of minor comments

File: trieste/models/keras/utils.py (hunk 1 above)
Comment (hstojic) [reply trieste/models/keras/utils.py]: size of the dataset is not needed for building neural nets, only dimensionality info is needed - I assume here we will only deal with usual matrices, thats why only the last dimension - I can make it more general

File: trieste/models/keras/utils.py (hunk 2 above)
Comment (hstojic) [reply trieste/models/keras/utils.py]: yeah, I wasn't thinking rank > 2 cases would be relevant, I'll do that

File: tests/unit/models/keras/test_utils.py (hunk 3 above)
Comment (hstojic) [reply tests/unit/models/keras/test_utils.py]: yep, will do

File: trieste/models/keras/utils.py (hunk 1 above)
Comment (hstojic) [reply trieste/models/keras/utils.py]: done, but I'm not sure `Dataset` would currently support that, I think it has an incompatible check at https://github.com/secondmind-labs/trieste/blob/6b2bb0e73649debaac81157f0f9fdb8d3fdfef5b/trieste/data.py#L51

**Pull request merged**